src/bvh/bvh.rs

```diff
@@ -163,30 +202,44 @@ impl BVHNode {
             child_r_indices.append(&mut indices);
         }
 
+        let node_index = nodes.len();
+
+        nodes.push(BVHNode::Dummy);
+
+        let child_l = BVHNode::build(shapes, child_l_indices, nodes, node_index, depth + 1);
+        let child_r = BVHNode::build(shapes, child_r_indices, nodes, node_index, depth + 1);
+
         // Construct the actual data structure
-        BVHNode::Node {
+        nodes[node_index] = BVHNode::Node {
+            parent: parent,
+            depth: depth,
             child_l_aabb: child_l_aabb,
-            child_l: Box::new(BVHNode::build(shapes, child_l_indices)),
+            child_l: child_l,
             child_r_aabb: child_r_aabb,
-            child_r: Box::new(BVHNode::build(shapes, child_r_indices)),
-        }
+            child_r: child_r,
+        };
+
+        node_index
     }
 
     /// Prints a textual representation of the recursive [`BVH`] structure.
     ///
     /// [`BVH`]: struct.BVH.html
     ///
-    fn pretty_print(&self, depth: usize) {
+    fn pretty_print(&self, nodes: &Vec<BVHNode>, depth: usize) {
```

Comment:
This method no longer conceptually fits into the implementation of `BVHNode`. It requires the whole tree, instead of just the node itself. It would make more sense to move this to `BVH`, what do you think?)